Add FactoryGirl/CreateList cop. Fixes #411

Darhazer opened the pull request:
Fixes #411 

create_list should be preferred over `n.times { :create }` (also over any loop that creates the same kind of object, but that's not implemented; n.times is the most common way I've encountered) as:
* it's provided by the FactoryGirl DSL
* it actually runs map, hence returning the created instances
* Just to enforce consistency

mikegee commented on lib/rubocop/cop/rspec/factory_girl/create_list.rb:
```diff
@@ -0,0 +1,67 @@
+          PATTERN
+
+          def_node_matcher :factory_call?, <<-PATTERN
+            (send nil :create (sym $_) $...)
```
Some apps I've worked on don't [include `FactoryGirl::Syntax::Methods `](https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#rspec) to enable this shorthand syntax and call `FactoryGirl.create` instead.

Darhazer commented:
PR updated

mikegee commented:
Should this cop enforce usage of the other [`*_list`](http://www.rubydoc.info/gems/factory_girl/FactoryGirl/Syntax/Methods) methods, too?

Darhazer commented:
I don't think anyone uses `build` inside of n.times as without reference to the built object you can't do anything useful. That's one difference between using n.times and create_list - the later returns the created objects

mikegee commented:
Good point

Darhazer commented:
Rebased on master

Darhazer commented:
@bquorning can you review this?

bquorning commented:
If you add a PR description I will :-)

bquorning commented on config/default.yml:
```diff
@@ -6,6 +6,7 @@ AllCops:
     - "(?:^|/)spec/"
   RSpec/FactoryGirl:
     Patterns:
+    - _spec.rb
```
What does this do?

Darhazer replied:
The first FactoryGirl cop checked just the factories, so it ignored the actual spec files. I need it to test them.

Darhazer replied:
Actually seems unneeded

bquorning commented on lib/rubocop/cop/rspec/factory_girl/create_list.rb:
```diff
@@ -0,0 +1,72 @@
+
+          def_node_matcher :n_times, '(send (int $_) :times)'
+
+          def_node_matcher :times_block?, <<-PATTERN
```
This matcher isn’t being used?

Darhazer replied:
Nice catch. I'll add a test for this

bquorning commented on lib/rubocop/cop/rspec/factory_girl/create_list.rb:
```diff
@@ -0,0 +1,72 @@
+          PATTERN
+
+          def on_block(node)
+            receiver, args, body = *node
```
Should we check that this is a `times` block? I guess that’s what `times_block?` was meant for…

Darhazer commented:
PR updated

bquorning commented on lib/rubocop/cop/rspec/factory_girl/create_list.rb:
```diff
@@ -0,0 +1,74 @@
+          def_node_matcher :n_times, '(send (int $_) :times)'
+
+          def_node_matcher :times_block_without_args?, <<-PATTERN
+          (block
```
Indentation is missing. Isn’t `Layout/IndentHeredoc` supposed to catch this?

Darhazer commented:
> Besides create, shouldn’t we enforce the same style for build, build_stubbed and attributes_for?

Only create could be used in this way in a loop. For the rest you need to store the result of a method somewhere, otherwise it just doesn't have a side effect. It's possible that someone does this in a loop and uses a variable, like `3.times { data << attributes_for(:user) }` but those are much harder to catch.

nevertheless, thanks for pointing the documentation again, I will add some tests for create used with a block.

bquorning commented:
How about e.g. `3.times.map { build(:user) }`?

Darhazer commented:
Wow, I've never seen such code.

Thanks for the feedback @bquorning , I will work on improving the cop in the next couple of weeks. It will need a betetr name as well, as soon as it's not only for create_list.

bquorning commented:
> Wow, I've never seen such code.

Neither have I 😃  And I don’t know if we should cover every possible case in this PR, or if we should wait and see if our users report an actual need. Probably the latter is fine.

backus commented:
Also I'm pretty sure `n.times.map { x }` is already autocorrected to `Array.new(n) { x }` by rubocop

backus commented:
Also I'm pretty sure `n.times.map { x }` is already autocorrected to `Array.new(n) { x }` by rubocop

bquorning commented on lib/rubocop/cop/rspec/factory_girl/create_list.rb:
```diff
@@ -0,0 +1,74 @@
+          )
+          PATTERN
+
+          def_node_matcher :factory_call?, <<-PATTERN
```
Since this matcher returns three captures (receiver, factory, and options) I would prefer we didn’t suffix it with `?` – that should be reserved for methods returning true/false.

bquorning commented on lib/rubocop/cop/rspec/factory_girl/create_list.rb:
```diff
@@ -0,0 +1,74 @@
+
+          def factory_call_replacement(body, count)
+            receiver, factory, options = *factory_call?(body)
+            replacement = receiver ? "#{receiver.source}." : ''
```
I would prefer to split this in two lines:
```rb
replacement = ''
replacement += "#{receiver.source}." if receiver
```

bquorning commented on lib/rubocop/cop/rspec/factory_girl/create_list.rb:
```diff
@@ -0,0 +1,74 @@
+            if options.count > 0
+              additional_options = options.map(&:source).join(', ')
+              replacement += ", #{additional_options}"
+            end
```
It seems this block could be simplified a bit:
```rb
options.each do |option|
  replacement += ", #{options.source}"
end
```

bquorning commented on spec/rubocop/cop/rspec/factory_girl/create_list_spec.rb:
```diff
@@ -0,0 +1,64 @@
+      used_passwords = []
+      3.times do
+        u = create :user
+        expect(userd_passwords).not_to include(u.password)
```
Typo – should be `used_passwords`.

Darhazer commented:
PR updated:
* Rebased on master
* Addessed bquorning comments, except for using just one matcher
* Added tests for possible false positive, if FactoryGirl.create is used with a block
* Preserved parens if the original code uses ones

Drenmi commented on lib/rubocop/cop/rspec/factory_girl/create_list.rb:
```diff
@@ -0,0 +1,92 @@
+          PATTERN
+
+          def on_block(node)
+            return unless times_block_without_args?(node)
```
Technically with arguments would also be an offense? Since there is a sequence helper in FactoryGirl. It just can't be easily auto-corrected. 🙂

Drenmi commented on lib/rubocop/cop/rspec/factory_girl/create_list.rb:
```diff
@@ -0,0 +1,92 @@
+            factory_call_replacement(body, count)
+          end
+
+          def method_uses_parens?(node)
```
The node extension for parameterised nodes has a `#parenthesized?` method. 🙂

walf443 commented:
FactoryGirl renamed to FactoryBot https://github.com/thoughtbot/factory_bot

walf443 commented:
I created  a PR that rename to FactoryBot https://github.com/backus/rubocop-rspec/pull/490

pirj commented on spec/rubocop/cop/rspec/factory_girl/create_list_spec.rb:
```diff
@@ -0,0 +1,76 @@
+
+  it 'flags usage of n.times when FactgoryGirl.create is used' do
+    expect_offense(<<-RUBY)
+      3.times { FactoryGirl.create :user }
```
FactoryBot as well.

pirj commented on spec/rubocop/cop/rspec/factory_girl/create_list_spec.rb:
```diff
@@ -0,0 +1,76 @@
+  it 'ignores FactoryGirl.create calls with a block' do
+    expect_no_offenses(<<-RUBY)
+      3.times do
+        create(:user) { |user| create :account, user: user }
```
But why? It could be `create_list(:user, 3) { |user| create :account, user: user }`

Darhazer replied:
The documentation doesn't mention this

pirj replied:
It's [in the spec](https://github.com/thoughtbot/factory_bot/blob/c716ce01b448ce4e0bf855c5a2c63ecb9206322e/spec/acceptance/create_list_spec.rb#L45). Tried in the console, works as well.

Darhazer replied:
10x. Will handle it

bquorning commented on CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@
 ## 1.19.0 (2017-10-18)
 
 Compatibility release so users can upgrade RuboCop to 0.51.0. No new features.
+* Add `FactoryGirl/CreateList` cop. ([@Darhazer][])
```
Please move this line further up in the Changelog :-)

bquorning commented on lib/rubocop/cop/rspec/factory_bot/create_list.rb:
```diff
@@ -0,0 +1,92 @@
+          PATTERN
+
+          def_node_matcher :factory_call, <<-PATTERN
+            (send ${(const nil? {:FactoryGirl :FactoryBot}) nil?} :create (sym $_) $...)
```
Couldn’t `{:FactoryGirl :FactoryBot}` be extracted to somewhere in lib/rubocop/rspec somehow?

bquorning commented on spec/rubocop/cop/rspec/factory_bot/create_list_spec.rb:
```diff
@@ -0,0 +1,83 @@
+    RUBY
+  end
+
+  it 'flags usage of n.times when FactgoryGirl.create is used' do
```
`s/Factgory/Factory/g` – 2 places needs fixing.

bquorning commented on lib/rubocop/cop/rspec/factory_bot/create_list.rb:
```diff
@@ -0,0 +1,92 @@
+            factory_call_replacement(body, count)
+          end
+
+          def method_uses_parens?(node)
```
Since it’s only called from `#format_method_call`, I’d prefer to have this method definition at the bottom of the class.

Darhazer replied:
No longer used

bquorning commented on lib/rubocop/cop/rspec/factory_bot/create_list.rb:
```diff
@@ -0,0 +1,92 @@
+          def_node_matcher :times_block_without_args?, <<-PATTERN
+            (block
+              #n_times
+              (args)
```
Help me: does `args` have a special meaning here?

Darhazer replied:
means no arguments passed to the blog (otherwise (args will be followed by the args yielded)

bquorning replied:
I’m asking because I don’t see it documented in https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/node_pattern.rb#L5-L91

Darhazer replied:
I'm just using `ruby-parse` to generate the ASTs. `(args) is used in some other cops too

Darhazer replied:
No longer needed, as now n.times with arguments is also flagged

Darhazer commented:
PR updated:
* Rebased on master
* Made the cop configurable, since create_list is slower than n.times and some users may prefer to enforce this style

Darhazer commented:
TODO: handle create_list with block

Darhazer commented:
Rebased and added handling for create with block

bquorning commented on lib/rubocop/cop/rspec/factory_bot/create_list.rb:
```diff
@@ -0,0 +1,154 @@
+
+            return unless n_times_block?(node)
+
+            return unless contains_only_factory?(node.body)
```
Nitpick: I prefer no empty lines between these three guard clauses.

bquorning commented on lib/rubocop/cop/rspec/factory_bot/create_list.rb:
```diff
@@ -0,0 +1,154 @@
+            end
+          end
+
+          def generate_replacement(block)
```
I would be fine with inlining this method into `#autocorrect_n_times_to_create_list`.

bquorning commented on lib/rubocop/cop/rspec/factory_bot/create_list.rb:
```diff
@@ -0,0 +1,154 @@
+          end
+
+          def build_options_string(options)
+            options.map { |option| ", #{option.source}" }.join
```
That leading comma feels a bit wrong here. How about e.g. `options.map(&:source).join(', ')`, and then something like `arguments += ', ' if options.any?` in the callers?

bquorning commented on lib/rubocop/cop/rspec/factory_bot/create_list.rb:
```diff
@@ -0,0 +1,154 @@
+          PATTERN
+
+          def_node_matcher :factory_list_call, <<-PATTERN
+            (send ${(const nil? {:FactoryGirl :FactoryBot}) nil?} :create_list (sym $_) $...)
```
…or even `(sym $_) (int $_) $...`? Then `count.source` would just become `count`.

Drenmi commented:
Thank you for working on this @Darhazer. 🙇
